src/ApacheOrcDotNet.OptimizedReader/Buffers/BaseColumnBuffer.cs

```diff
@@ -234,71 +176,86 @@ private protected void ReadNumericStream(StreamDetail stream, ReadOnlySpan<byte>
         {
             numValuesRead = 0;
 
             if (stream == null)
                 return;
 
             int numSkipped = 0;
-            var bufferReader = new BufferReader(GetDataStream(stream, buffer, length));
+            var bufferReader = new BufferReader(GetDataStream(stream, decompressedBuffer));
 
             while (!bufferReader.Complete)
             {
-                var bigInt = ReadVarInt(ref bufferReader);
+                var varInt = ReadVarInt(ref bufferReader);
 
-                if (!bigInt.HasValue)
+                if (!varInt.HasValue)
                     return;
 
                 if (numSkipped++ < stream.Positions.ValuesToSkip)
                     continue;
 
-                outputValues[numValuesRead++] = bigInt.Value;
+                outputValues[numValuesRead++] = varInt.Value;
 
                 if (numValuesRead >= outputValues.Length)
                     return;
             }
         }
 
-        private protected async Task<int> GetByteRangeAsync(StreamDetail stream, Memory<byte> output)
+        private protected static void CheckByteRangeBufferLength(StreamDetail stream, ref byte[] targetBuffer)
         {
-            if (stream != null)
-            {
-                // If current and last ranges are equal, the previous data will be buffered
-                // and we can return only the length, without requesting the bytes again.
+            if (stream == null)
+                return;
 
-                if (stream.Range != _lastRange)
-                    _lastRangeLength = await _byteRangeProvider.GetRangeAsync(output.Slice(0, stream.Range.Length), stream.Range.Offset);
+            if (stream.Range.Length > targetBuffer.Length)
+                targetBuffer = new byte[stream.Range.Length];
+        }
 
-                _lastRange = stream.Range;
-            }
+        private protected async Task<int> GetByteRangeAsync(StreamDetail stream, Memory<byte> outputBuffer)
+        {
+            if (stream == null)
+                return 0;
+
+            // If current and last ranges are equal, the previous data will be buffered
+            // and we can return only the length, without requesting the bytes again.
+
+            if (stream.Range == _lastRange)
+                return _lastRangeLength;
+
+            await _byteRangeProvider.FillBufferAsync(outputBuffer[..stream.Range.Length], stream.Range.Offset);
+
+            _lastRangeLength = stream.Range.Length;
+            _lastRange = stream.Range;
 
             return _lastRangeLength;
         }
 
-        private protected void DecompressByteRange(StreamDetail stream, ReadOnlySpan<byte> compressedInput, Span<byte> decompressedOutput, ref int decompressedLength)
+        private protected void DecompressByteRange(StreamDetail stream, ReadOnlySpan<byte> compressedInput, ref byte[] decompressedOutput, ref int decompressedLength)
```

Comment:
Crazy question: you don't want to just return a reference to a thread-local static buffer from CompressedData.Decompress do you?  Let that call manage all of this?  We could even move back to the lambda provided to the decompression routine that hands off the buffer and then expects it back at the end of the `Action`...

Reply:
Based on our last call, this will be considered in a future improvement (`CompressData` would need to know about all possible buffers per stream and data types).